jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1283 @@
+ * object. Therefore, calling the remove methods concurrently with a typical algorithm is likely to
+ * cause the algorithm to fail with an {@link IllegalArgumentException}. So really the main
+ * concurrent read/write use case is add-only.
+ * <br>
+ * eg: If threadA tries to get all edges touching a certain vertex after threadB removes the vertex,
+ * the algorithm will be interrupted by {@link IllegalArgumentException}.
+ * </p>
+ * <pre>
+ *      Thread threadA = new Thread(() -&gt; {
+ *          Set vertices = graph.vertexSet();
+ *          for (Object v : vertices) {
+ *              // {@link IllegalArgumentException} may be thrown since other threads may have removed the vertex.
+ *              Set edges = graph.edgesOf(v);
+ *              doOtherThings();
+ *          }
+ *      });
+ *      Thread threadB = new Thread(() -&gt; {
+ *          Set vertices = graph.vertexSet();
+ *          for (Object v : vertices) {
+ *              if (someConditions)
+ *                  graph.removeVertex(v);
+ *          }
+ *      });
+ * </pre>
+ *
+ * <p>
+ * The created Graph's hashCode is equal to the backing set's hashCode. And the created Graph is equal
+ * to another Graph if they are the same Graph or the backing Graph is equal to the other Graph.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author CHEN Kui
+ * @since Feb 23, 2018
+ */
+public class AsSynchronizedGraph<V, E>
+    extends GraphDelegator<V, E>
+    implements Graph<V, E>, Serializable
+{
+    private static final long serialVersionUID = 5144561442831050752L;
+
+    private final ReentrantReadWriteLock readWriteLock;
+
+    // A set encapsulating backing vertexSet.
+    private transient CopyOnDemandSet<V> allVerticesSet;
+
+    // A set encapsulating backing edgeSet.
+    private transient CopyOnDemandSet<E> allEdgesSet;
+
+    private CacheStrategy<V, E> cacheStrategy;
+
+    /**
+     * Constructor for AsSynchronizedGraph with strategy of not caching the copies for
+     * <code>edgesOf</code>, <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods
+     * and non-fair mode for thread-access.
+     *
+     * @param g the backing graph (the delegate)
+     */
+    public AsSynchronizedGraph(Graph<V, E> g)
+    {
+        this(g, false, false);
+    }
+
+    /**
+     * Constructor for AsSynchronizedGraph with specified cache strategy for <code>edgesOf</code>,
+     * <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods and specified fairness
+     * policy for thread-access.
+     *
+     * @param g the backing graph (the delegate)
+     * @param cacheEnabled a flag whether to use cache for those methods, if <tt>true</tt>, cache
+     *        will be used for those methods, otherwise cache will not be used
+     * @param fair fairness policy for thread-access
+     */
+    private AsSynchronizedGraph(Graph<V, E> g, boolean cacheEnabled, boolean fair)
```

Comment:
I think we should get rid of this constructor and require the user to call the version which takes a SynchronizedGraphParams.